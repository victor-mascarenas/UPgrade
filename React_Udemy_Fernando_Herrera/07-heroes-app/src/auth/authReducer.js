/* const state = {
    name: 'Victor',
    logged: true
}; */

import { types } from "../types/types";

export const authReducer = (state = {}, action) => {
    switch(action.type) {
        case types.login:
            state = {
                name: action.payload.name,
                logged: true
            };
            break;
        case types.logout:
            state = {
                logged: false
            }
            break;
        default:
            break;
    }
    return state;
};