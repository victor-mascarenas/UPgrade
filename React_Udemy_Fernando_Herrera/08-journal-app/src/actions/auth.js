import { types } from "../reducers/types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { endLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return (dispatch) => {//dispatch comes from Thunk
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch(error => {
                Swal.fire('Error', error.message, 'error');
            })
            .finally(() => {
                dispatch(endLoading());
            });
    };
};

export const startCredentialsRegistration = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({user}) => {
                await user.updateProfile({
                    displayName: name
                });
                dispatch(login(user.uid, user.displayName));
            })
            .catch(error => {
                Swal.fire('Error', error.message, 'error');
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName));
            })
            .catch((error) => {
                Swal.fire('Error', error.message, 'error');
            });
    };
};

export const login = (uid, displayName) => (
    {
        type: types.LOGIN,
        payload: {
            uid,
            displayName
        }
    }
);

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout());
    };
};

export const logout = () => {
    return {
        type: types.LOGOUT
    };
}