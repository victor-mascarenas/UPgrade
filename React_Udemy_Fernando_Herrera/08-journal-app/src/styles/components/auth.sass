.auth__main
    align-items: center
    background-color: $primary
    display: flex
    justify-content: center
    height: 100vh
    margin: 0px
    width: 100vw
.auth__box-container
    background-color: white
    border-radius: 2px
    box-shadow: 0px 3px $dark-gray
    padding: 20px
    width: 250px