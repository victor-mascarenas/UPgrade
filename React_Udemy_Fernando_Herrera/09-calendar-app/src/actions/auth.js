import Swal from "sweetalert2";
import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

const storeToken = (token) => {
    localStorage.setItem('token', token);
    localStorage.setItem('token-start', new Date().getTime());
};

export const startLogin = (email, password) => {
    return async (dispatch) => {
        const res = await fetchWithoutToken('auth', {
            email,
            password
        }, 'POST');
        const body = await res.json();
        if (body.ok) {
            storeToken(body.token);
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    };
};
export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        const res = await fetchWithoutToken('auth/new', {
            name,
            email,
            password
        }, 'POST');
        const body = await res.json();
        if (body.ok) {
            storeToken(body.token);
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    };
};

const login = (user) => ({
    type: types.AUTH_LOGIN,
    payload: user
});