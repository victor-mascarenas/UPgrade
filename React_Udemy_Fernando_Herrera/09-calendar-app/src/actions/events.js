import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types"

const eventAddNew = (event) => ({
    type: types.EVENT_ADD_NEW,
    payload: event
});
const eventLoaded = (events) => ({
    type: types.EVENT_LOADED,
    payload: events
});

export const eventStartAddNew = (event) => {
    return async (dispatch, getState) => {
        const {uid, name} = getState().auth;
        try {
            const resp = await fetchWithToken('events', event, 'POST');
            const body = await resp.json();
            if (body.ok) {
                event.id = body.evento.id;
                event.user = {
                    _id: uid,
                    name
                };
                dispatch(eventAddNew(event));
            } else {
                Swal.fire('Error', body.msg, 'error');
            }
        } catch(error) {
            Swal.fire('Error', error.message, 'error');
        }
    };
};
export const eventStartLoading = () => {
    return async (dispatch) => {
        try {
            const res = await fetchWithToken('events');
            const body = await res.json();
            if (body.ok) {
                const events = prepareEvents(body.events);
                dispatch(eventLoaded(events));
            } else {
                Swal.fire('Error', body.msg, 'error');
            }
        } catch (error) {
            Swal.fire('Error', error.message, 'error');
        }
    };
};

export const eventSetActive = (event) => ({
    type: types.EVENT_SET_ACTIVE,
    payload: event
});
export const eventClearActive = () => ({
    type: types.EVENT_CLEAR_ACTIVE
});
export const eventUpdated = (event) => ({
    type: types.EVENT_UPDATED,
    payload: event
});
export const eventDeleted = () => ({
    type: types.EVENT_DELETED
});
export const eventSlotSelected = (start, end) => ({
    type: types.EVENT_SLOT_SELECTED,
    payload: {
        start,
        end
    }
});