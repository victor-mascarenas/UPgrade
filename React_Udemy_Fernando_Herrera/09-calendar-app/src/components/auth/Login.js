import React from 'react'

export const Login = () => {
    return (
        <div>
            <h1>Login Screen</h1>
        </div>
    )
};
