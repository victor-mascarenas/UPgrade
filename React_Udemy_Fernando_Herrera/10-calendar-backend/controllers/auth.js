const bcrypt = require("bcryptjs/dist/bcrypt");
const {response} = require("express");
const { encrypt } = require("../helpers/crypto");
const { generateGeneric } = require("../helpers/responseGenerator");
const { generateJWT } = require("../helpers/tokens");
const User = require("../models/User");

const isEmailRelatedToAnyUser = async (email) => {
    let related = false;
    const dbUser = await User.findOne({
        email
    });
    if (dbUser) {
        related = true;
    }
    return related;
};

const createUser = async (req, res = response) => {
    let resp;
    const {email, password} = req.body;
    try {
        if (await isEmailRelatedToAnyUser(email)) {
            resp = generateGeneric(res, 400, 'El correo ya esta asignado a un usuario');
        } else {
            const user = new User(req.body);
            user.password = encrypt(password);
            await user.save();
            const token = await generateJWT(user.id, user.name);
            resp = res.status(201).json({
                ok: true,
                uid: user.id,
                name: user.name,
                token
            });
        }
    } catch (error) {
        resp = generateGeneric(res, 500, 'Por favor contacte al administrador');
    }
    return resp;
};
const userLogin = async (req, res = response) => {
    let resp;
    const {email, password} = req.body;
    try {
        const dbUser = await User.findOne({
            email
        });
        if (!dbUser) {
            resp = generateGeneric(res, 400, 'El usuario no existe con ese email');
        } else {
            const validPassword = bcrypt.compareSync(password, dbUser.password);
            if (!validPassword) {
                resp = generateGeneric(res, 400, 'ContraseÃ±a invalida');
            } else {
                const token = await generateJWT(dbUser.id, dbUser.name);
                resp = res.status(201).json({
                    ok: true,
                    uid: dbUser.id,
                    name: dbUser.name,
                    token
                });
            }
        }
    } catch (error) {
        resp = generateGeneric(res, 500, 'Por favor contacte al administrador');
    }
    return resp;
};
const revalidateToken = async (req, res = response) => {
    let resp;
    const {uid, name} = req;
    try {
        const token = await generateJWT(uid, name);
        resp = res.status(200).json({
            ok: true,
            uid: uid,
            name: name,
            token
        });
    } catch (error) {
        resp = generateGeneric(res, 500, 'Por favor contacta al administrador');
    }
    return resp;
};

module.exports = {
    createUser,
    userLogin,
    revalidateToken
};