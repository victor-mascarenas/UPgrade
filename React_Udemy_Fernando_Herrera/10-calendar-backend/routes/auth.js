const {Router} = require('express');
const {createUser, userLogin, revalidateToken} = require('../controllers/auth');
const {check} = require('express-validator');
const validateFields = require('../middlewares/field-validator');
const { validateJWT } = require('../middlewares/token-validator');

const router = Router();

/*
...post([route], [middlewares], [controller function]);
*/
router.post('/new', [
    check('name', 'El nombre es requerido').not().isEmpty(),
    check('email', 'El email no es valido').isEmail(),
    check('password', 'La contraseña debe contener al menos 6 caracteres').isLength({
        min: 6
    }),
    validateFields
], createUser);
router.post('/', [
    check('email', 'El email no es valido').isEmail(),
    check('password', 'La contraseña debe contener al menos 6 caracteres').isLength({
        min: 6
    }),
    validateFields
], userLogin);
router.get('/renew', [
    validateJWT
], revalidateToken);

module.exports = router;