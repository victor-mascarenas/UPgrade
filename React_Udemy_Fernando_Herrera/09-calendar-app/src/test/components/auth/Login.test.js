import React from 'react';
import {mount} from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Login } from '../../../components/auth/Login';
import { startLogin } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <Login/>
    </Provider>
);

describe('Pruebas en Login', () => {
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('Debe de llamar el dispatch del login', () => {
        const email = 'juan@gmail.com';
        const password = '123456';
        wrapper.find('input[name="loginEmail"]').simulate('change', {
            target: {
                name: 'loginEmail',
                value: email
            }
        });
        wrapper.find('input[name="loginPassword"]').simulate('change', {
            target: {
                name: 'loginPassword',
                value: password
            }
        });
        wrapper.find('form').at(0).prop('onSubmit') ({
            preventDefault: jest.fn()
        });
        expect(startLogin).toHaveBeenCalledWith(email, password);
    });
});