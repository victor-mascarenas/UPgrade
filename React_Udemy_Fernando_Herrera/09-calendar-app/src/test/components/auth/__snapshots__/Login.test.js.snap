// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Login Debe mostrarse correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit btn"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit btn"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
